### PR TITLE
merge: Avoid overwriting the output id column with non-id columns

### PR DESCRIPTION
**Pull request by tsibley**
This edge case occurs when an input table has an id column used for joining (in tests, literally "id", but any valid id column name) and also another, non-id column that matches the output id column name (in tests, literally "strain").

## Checklist

- [x] Automated checks pass
- [x] [Check][1] if you need to add a changelog message
- [x] [Check][2] if you need to add tests
- [x] [Check][3] if you need to update docs

[1]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#updating-the-changelog
[2]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#testing
[3]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#when-to-update

### CODE REVIEW

File: augur/merge.py [hunk 1]
```diff
@@ -186,7 +188,24 @@ def run(args):
...
+                # Don't overwrite output id column (i.e. first table's id
+                # column) with a non-id column of the same name (i.e. from a
+                # subsequent table).
+                #
+                # XXX TODO: Alternatively, we could a) skip such columns rather
+                # than mangle their names, or b) allow for output columns with
+                # non-unique names (i.e. distinguishable only by position).
+                # Would either of those be preferrable?  Should it be
+                # configurable?
+                #   -trs, 21 Aug 2024
```
Comment (tsibley) [inline augur/merge.py:203]: I'd welcome commentary on these alternatives.

Comment (victorlin) [reply augur/merge.py:203]: I would avoid (b) since this could cause complications downstream. I'm guessing many workflows assume unique column names. Other Augur commands that use `read_metadata` seem to allow it since [pandas allows it](https://github.com/pandas-dev/pandas/issues/47718), but this seems more like a bug waiting to break some workflow rather than an intentional feature.

Decision here could benefit from having an example use case. Would it be unreasonable to (c) exit with error? This would allow us to better detect if/when this ever happens, and figure out what seems best based on the issue at hand.

Comment (genehack) [reply augur/merge.py:203]: I don't have a strong opinion about the two alternatives, but I do think, if you're going to mangle the column name, you need to consider the edgier edge case of "what if we mangle it to something that already exists?" (which will also cover the edge case of "what if there is more than one non-id column whose name overlaps the id column", I think).

Comment (joverlee521) [reply augur/merge.py:203]: (c) exit with error + suggest `augur curate rename` for users to rename columns?

Comment (tsibley) [reply augur/merge.py:203]: @genehack 

> but I do think, if you're going to mangle the column name, you need to consider the edgier edge case of "what if we mangle it to something that already exists?"

Yeah, my prefixing with two underscores approach is punting on this under the rationale that "field names with two leading underscores are reserved for use by augur merge". I didn't spell this out in docs, but it's kinda implied given this behaviour and the pre-existing `__source_metadata_X` columns.

> (which will also cover the edge case of "what if there is more than one non-id column whose name overlaps the id column", I think).

Hmm, are you thinking of a situation of two tables:

```csv
strain,…
1,…
2,…
3,…
```

```csv
id,strain,strain,…
1,A,B,…
2,A,B,…
3,A,B,…
```

which are matched on `strain` in table 1 and `id` in table 2?

Maybe we should support that situation somehow, but currently we don't and it's a error/bug:

```console
$ augur merge --metadata a=/tmp/a.csv b=/tmp/b.csv --output-metadata -
Reading 'a' metadata from '/tmp/a.csv'…
Reading 'b' metadata from '/tmp/b.csv'…
Columns renamed during .import <pipe> due to duplicates:
"strain" to "strain_2",
"strain" to "strain_3"
Error: in prepare, no such column: "strain" - should this be a string literal in single-quotes?
  eate unique index "metadata_b_id" on "metadata_b"("strain");
                                      error here ---^
WARNING: Skipped deletion of /tmp/augur-merge-ejdrz56t.sqlite due to error, but you may want to clean it up yourself (e.g. if it's large).
ERROR: sqlite3 invocation failed
```

Comment (tsibley) [reply augur/merge.py:203]: While I like the implementation simplicity of (c) exit with error, I don't like where it leaves the user. I'd rather it Just Work and leave the user with the option to continue on uninterrupted with the mangled name or rename it to their preferred name after the merge.

But maybe mangling is less friendly if users are likely to not notice it happened, even with the warning message?

Seems like best thing is to pick one way and get a sense from actual usage.

Comment (victorlin) [reply augur/merge.py:203]: > But maybe mangling is less friendly if users are likely to not notice it happened, even with the warning message?

I'm no user so take this with a grain of salt, but this seems likely in the context of Snakemake usage where warnings are effectively log messages that get drowned out in the sea of console output.

Comment (genehack) [reply augur/merge.py:203]: > Yeah, my prefixing with two underscores approach is punting on this under the rationale that "field names with two leading underscores are reserved for use by augur merge". I didn't spell this out in docs, but it's kinda implied given this behaviour and the pre-existing `__source_metadata_X` columns.

If that's going to be the behavior, I think that's fine, but (a) it probably should be documented and (b) you should probably throw warnings when you see `__<column_name>` in inputs, yeah? 

> > (which will also cover the edge case of "what if there is more than one non-id column whose name overlaps the id column", I think).
> 
> Hmm, are you thinking of a situation of two tables:

I was thinking about merging three or more tables, one where `strain` is the unique ID column and two where it is not but is still present. 

> Maybe we should support that situation somehow


, but currently we don't and it's a error/bug:
> 
> ```
> $ augur merge --metadata a=/tmp/a.csv b=/tmp/b.csv --metadata-id-columns a=strain b=id --output-metadata -
> Reading 'a' metadata from '/tmp/a.csv'…
> Reading 'b' metadata from '/tmp/b.csv'…
> Columns renamed during .import <pipe> due to duplicates:
> "strain" to "strain_2",
> "strain" to "strain_3"
> Traceback (most recent call last):
>   File "/home/tom/nextstrain/augur/augur/__init__.py", line 70, in run
>     return args.__command__.run(args)
>   File "/home/tom/nextstrain/augur/augur/merge.py", line 209, in run
>     assert m.columns == (table_columns := sqlite3_table_columns(db_path, m.table_name)), \
> AssertionError: ['id', 'strain', 'strain', '…'] == ['id', 'strain_2', 'strain_3', '…']
> 
> 
> An error occurred (see above) that has not been properly handled by Augur.
> To report this, please open a new issue including the original command and the error above:
>     <https://github.com/nextstrain/augur/issues/new/choose>
> ```

Comment (tsibley) [reply augur/merge.py:203]: I'm going to switch this PR's implementation to the suggested (c) exit with error. I suspect we'll encounter it more than we'd like—that is, I suspect we have plenty of metadata files to merge in where, e.g., `strain` is some non-unique column and we're joining on `accession` or similar instead—but we can address it down the road if so.

> If that's going to be the behavior, I think that's fine, but (a) it probably should be documented and (b) you should probably throw warnings when you see `__<column_name>` in inputs, yeah?

Fair enough.

> I was thinking about merging three or more tables, one where `strain` is the unique ID column and two where it is not but is still present.

Ah. In the current (but soon to be outdated) implementation on this PR, that'd be fine; they'd both get renamed in the output to `__strain` and merged appropriately. Or is there something else I'm missing?

Comment (genehack) [reply augur/merge.py]: > > I was thinking about merging three or more tables, one where `strain` is the unique ID column and two where it is not but is still present.
> 
> Ah. In the current (but soon to be outdated) implementation on this PR, that'd be fine; they'd both get renamed in the output to `__strain` and merged appropriately. Or is there something else I'm missing?

Perhaps I misunderstood the desired end state — I thought the goal was that the final output of `augur merge` would have fully unique column names. If the goal is actually "fully unique except `__`-prefixed ones", then 👍 👍

Comment (tsibley) [reply augur/merge.py]: You understand the desired end state. The output would have fully-unique column names. The two `__strain` columns would get merged, i.e. output as a single column.

**Comment by tsibley** [comment]
The linkcheck step in CI has now failed twice in different ways: first reporting a 403 Forbidden for <http://www.microbesonline.org/fasttree/> and then reporting 406 Unacceptable for <https://github.com/nextstrain/augur/commit/4859b5d70e77cc9a0bb99e741fefb29952058b71>. Third time was the charm. The only thing I did was re-run the job.

**Review by victorlin — APPROVED** [review]
